Benke lukas

benkelukas opened the pull request:
- Added experimental docker - maybe someone can use it :)
- Added ValueObjects, Entities, Interfaces and Services
- Added tests
- Updated README.md to include instructions on running this inside docker

Commit:
updated gitignore

Commit:
docker done

Commit:
SSH for docker

Commit:
basic stuff done

Commit:
reformatted source code according to PSR2

Commit:
Native Number type - pre fix commit

Commit:
Added exceptions to IssuedInvoice constructor
Refactor checkByIco to isPresentByIco for better clarity
Added Nullable ICO VO to Buyer object
Added some tests to back it up

Commit:
Added exceptions to IssuedInvoice constructor
Refactor checkByIco to isPresentByIco for better clarity
Added Nullable ICO VO to Buyer object
Added some tests to back it up

Commit:
Added exceptions to IssuedInvoice constructor
Refactor checkByIco to isPresentByIco for better clarity
Added Nullable ICO VO to Buyer object
Added some tests to back it up

Commit:
updated readme with info about docker

Commit:
removed Notes.md from repo

Commit:
added void return type to tests

Commit:
updated readme with some more info regarding docker

simara-svatopluk commented on src/Domain/ValueObjects/PaymentInformation.php:
```diff
@@ -0,0 +1,30 @@
+
+namespace DDDWorkshop\Domain\ValueObjects;
+
+class PaymentInformation
```
This was meant to be combination of "bank account number" and "variable symbol"

benkelukas replied:
Implemented. Thanks

simara-svatopluk commented on tests/Domain/Entities/IssuedInvoiceTest.php:
```diff
@@ -0,0 +1,152 @@
+    /**
+     * @test
+     */
+    public function itCanCreateItselfWithProvidedInformation():void
```
what about name "canBeCreatedByProperInformation"?

benkelukas replied:
Yes, that is better :)

simara-svatopluk commented on tests/Domain/Entities/IssuedInvoiceTest.php:
```diff
@@ -0,0 +1,152 @@
+    public function itCanGetTotalAmount():void
+    {
+        /**
+         * TODO: Jak dostanu posledni dve veci do tridy, pokud je budu mit v DIC kontejneru nabindovane pres interface?
```
This will be discussed during 9th lecture - commands. Till that point we will wire it manually. If you don't want to wait, check https://pehapkari.cz/blog/2018/02/28/domain-driven-design-repository/ , section Persistence Responsibility, and the image with Handler. Handler is a service (application service, in DI container) that receives dependencies via constructor.

benkelukas replied:
Ok, thanks, will take a look at this when i get time

simara-svatopluk commented on tests/Domain/Entities/IssuedInvoiceTest.php:
```diff
@@ -0,0 +1,152 @@
+            new Series()
+        );
+
+        $this->assertEquals(1700, $issuedInvoice->getTotalAmount()->toFloat());
```
this test is resting responsibility of `InvoiceItems`. If the algorithm of `InvoiceItems::totalAmount()` changes, you'll have to change also this test.

But this test make sense, because it tests the most important Entity. So I suggest not comparing it to exact value, but rather `$this->assertGreaterThan(0, ...)`

benkelukas replied:
Well, i would rather check the underlying algorithm also. The algorithm is not injected into the class, so i think it is the responsibility of the test to check if the algorithm works correctly. That is why i think that it is OK for the test to change if algorithm changes. What do you think about that?

simara-svatopluk replied:
It could be a longer discussion, so please keep this topic in mind for next course.

When I test I want to test only one thing.

As you wrote `check the underlying algorithm also` this means that you want to check two things. And checking two things is something I don't want to do. And also as you wrote `test to change if algorithm changes` - this is exactly what I suggest to not test.

simara-svatopluk commented on tests/Domain/Entities/IssuedInvoiceTest.php:
```diff
@@ -0,0 +1,152 @@
+        $this->expectException(DateOfPaymentIsBeforeDateOfIssue::class);
+
+        new IssuedInvoice(
+            new Seller(
```
you have frequently repeated code in the test that is not relevant for test use cases. I suggest to extract pieces like
`new Seller....` into private functions `$this->createCorrectSeller()`

But not all of them, only those that are not relevant for test use case.... choose wisely

benkelukas replied:
Done, thanks

simara-svatopluk commented on src/Domain/ValueObjects/Date.php:
```diff
@@ -0,0 +1,138 @@
+use InvalidArgumentException;
+use function preg_match;
+
+class Date
```
Did you consider using some Date library instead developing own code? Because you have to also maintain your own code ;-)

benkelukas replied:
Yes i did. And in some fit of madness i did not do it, guess i wanted some exercise in the moment :) I have swapped my implementation for Carbon, that should be better.

simara-svatopluk commented on src/Domain/ValueObjects/Date.php:
```diff
@@ -0,0 +1,138 @@
+     *
+     * @return bool
+     */
+    public function laterThan(Date $object): bool
```
this function is never publicly, maybe it should be private

simara-svatopluk commented on src/Domain/ValueObjects/Date.php:
```diff
@@ -0,0 +1,138 @@
+        if (!preg_match("/^([1-9][1-9]*)\.([1-9][1-9]*)\.([0-9]{4})$/", $dateInCzechFormat)) {
+            throw new InvalidArgumentException('Invalid date given');
+        }
+        $dt = DateTimeImmutable::createFromFormat('j.n.Y', $dateInCzechFormat);
```
I have a suggestion for this entire class - use `DateTimeImmutable` as the internal state. It is maintained by someone else and if you use it, you can get rid of a lot of code

simara-svatopluk commented on src/Domain/Entities/IssuedInvoice.php:
```diff
@@ -0,0 +1,169 @@
+        //get next number in series
+        $number = $series->getNextNumber();
+        //check that date of issue is before date of payment
+        if (!$dateOfIssue->getDate()->earlierThan($dateOfPayment->getDate())) {
```
I think that this comparison could be responsibility of `DateOfIssue` or `DateOfPayment`

imagine that instead of this `if`, it would be there `$dateOfIssue->checkIsBefore($dateOfPayment); //throws DateOfPaymentIsBeforeDateOfIssue`
then you could even remove the comment above

simara-svatopluk replied:
One point to the logic - the `date of issue` have to be **less or equal** to `date of payment`

benkelukas replied:
Thanks, i have moved the check into guard methods on respective classes. Great suggestion :)

simara-svatopluk commented on src/Domain/Entities/IssuedInvoice.php:
```diff
@@ -0,0 +1,169 @@
+        if ($registrNespolehlivychPlatcu->isPresentByIco($seller->getIco())) {
+            throw new SellerIsOnRegistrNespolehlivychPlatcu($seller);
+        }
+        //get next number in series
```
please remove this comment, it is explained by the code itself ;-)

benkelukas replied:
Done

simara-svatopluk commented on src/Domain/Entities/IssuedInvoice.php:
```diff
@@ -0,0 +1,169 @@
+        SeriesInterface $series
+    ) {
+        //check Registr Nespolehlivych Platcu
+        if ($registrNespolehlivychPlatcu->isPresentByIco($seller->getIco())) {
```
you can move responsibility of checking into the `RegistrNespolehlivychPlatcuInterface`. change method `isPresentByIco` to `check`, and the check method will throw an exception if the ICO is in the register

benkelukas replied:
Same as DateOfIssue

simara-svatopluk commented on src/Domain/ValueObjects/InvoiceItem.php:
```diff
@@ -0,0 +1,66 @@
+
+namespace DDDWorkshop\Domain\ValueObjects;
+
+class InvoiceItem
```
be careful, we have only `Item` and `Items` in domain language ~(not `InvoiceItem`)~

simara-svatopluk commented on src/Domain/ValueObjects/InvoiceItems.php:
```diff
@@ -0,0 +1,88 @@
+    private $items;
+
+    /**
+     * InvoiceItems constructor.
```
I don't find these constructor comments useful

simara-svatopluk replied:
I don't find comments useful at all, because no one maintain them

benkelukas replied:
Those were autogenerated, sorry :)

simara-svatopluk commented on src/Domain/ValueObjects/InvoiceItems.php:
```diff
@@ -0,0 +1,88 @@
+     */
+    public function __construct(array $items = [])
+    {
+        $this->items = $this->validate($items);
```
It is strange for me that `validate` method returns something. `validate` method should only do it's job - validate. So I prefer

```php
$this->validate($items);
$this->items = $items; // or $this->assignItems if there is more logic, but it is not there
```

simara-svatopluk commented on src/Domain/ValueObjects/InvoiceItems.php:
```diff
@@ -0,0 +1,88 @@
+    /**
+     * @return int
+     */
+    public function count(): int
```
we don't have usage for this method.... do you really need it?

simara-svatopluk commented on src/Domain/ValueObjects/InvoiceItems.php:
```diff
@@ -0,0 +1,88 @@
+     */
+    public function totalAmount(): TotalAmount
+    {
+        $total = (float)array_reduce($this->items, function (?int $carry, InvoiceItem $item) {
```
great function. You can also provide a default value as a 3rd parameter of `array_reduce`, and then the typehint in `$carry` could be `int` without question mark.... but wait, it should be `float`

simara-svatopluk commented on src/Domain/Services/RegistrNespolehlivychPlatcu.php:
```diff
@@ -0,0 +1,16 @@
+use DDDWorkshop\Domain\Interfaces\RegistrNespolehlivychPlatcuInterface;
+use DDDWorkshop\Domain\ValueObjects\ICO;
+
+class RegistrNespolehlivychPlatcu implements RegistrNespolehlivychPlatcuInterface
```
Since these classes (together wih Series) are used only in tests, I recommend moving them to tests and renaming to something like `PassingRegistrNespolehlivychPlatcu` and `StaticSeries`

It is ok to have an interface in Domain layer without implementation

simara-svatopluk commented on src/Domain/ValueObjects/ICO.php:
```diff
@@ -0,0 +1,77 @@
+            throw new InvalidArgumentException('Invalid ICO pattern');
+        }
+
+        // checksum
```
I recommend to move next 5 lines into private method `checkChecksum`

simara-svatopluk commented on src/Domain/ValueObjects/ICO.php:
```diff
@@ -0,0 +1,77 @@
+
+class ICO
+{
+    protected $ico_pattern = '#^\d{8}$#';
```
private const would suite better

simara-svatopluk commented on tests/Domain/InvoiceItemsTest.php:
```diff
@@ -0,0 +1,77 @@
+use DDDWorkshop\Domain\ValueObjects\InvoiceItems;
+use PHPUnit\Framework\TestCase;
+
+class InvoiceItemsTest extends TestCase
```
you can also test immutability - that after adding an item, old items still return the same total amount and new items return a new total amount

simara-svatopluk commented on src/Domain/ValueObjects/NullICO.php:
```diff
@@ -0,0 +1,14 @@
+
+namespace DDDWorkshop\Domain\ValueObjects;
+
+class NullICO extends ICO
```
We didn't speak about `NullICO` with domain expert. What is the motivation for this new concept? What problems it solves? Is it necessary?

This is the only really problematic part of your code in my opinion (but it doesn't immediately mean it is wrong)

benkelukas replied:
Well, user can Issue invoices for someone who does not have ICO and is identified only by Name, At least that is what i remember from the lectures. But it is possible that my brain had made that up. If we are gonna issue invoices only to buyers with ICO, than i will remove it.

benkelukas replied:
Also, reason for namespaces is just my preference. I like having stuff separated and in my mind the VO, Entities, Services etc. are different enough to warrant a separate place to put them.

simara-svatopluk replied:
* `NullICO` - I thought that `null` could be suitable for these use cases... But in the other hand, returning `ICO` is much better that returning `?ICO`.... hm... I'll have to think about it more
* namespaces - ok, separation for easier thinking is fine

simara-svatopluk reviewed (commented):
All in all great job!

I have couple of questions and suggestions that are done for improvements, not for critics.

Why do you have Value Objects, Entities, Services, ... in different namespaces? What is the reason?

Commit:
Changed PaymentInformation VO to be composite of BankAccountNumber and VariableSymbol
Changed constructor tests to canBeCreatedByProperInformation for better clarity
Extracted test code that created some static VOs to methods
Switched from DIY Date class to Carbon
Made DateOfIssue VO responsible for enforcing DateOfPayment being later
than DateOfIssue
Removed unnecessary comments
Added check guard method to RegistrNespolehlivychPlatcu
Validate methods on VOs are now void
removed unused count() method from Items VO
Simplified counting of Items Total
Extracted ICO check checksum logic to method

Commit:
Added immutability test to Items collection

Commit:
moved series and registr nespolehlivych platcu services to test namespace
IssuedInvoice does not check for ICO on Registr Nespolehlivych Platcu
when ICO is NullICO

simara-svatopluk commented:
@benkelukas hi, I'm not able to answer now, but I'll do it in a week (most likely on Thursday)

benkelukas commented:
@simara-svatopluk Hi, no problem, whenever You have time. Thanks :)

simara-svatopluk commented:
Thanks for the PR, all in all it is great piece of code!

benkelukas commented:
Thanks for the review. It was very helpful and on-point.

simara-svatopluk commented:
The DDD course is over, closing
